REGRESSION(267321@main): ASSERTION FAILED: style() in WebCore::CSSToLengthConversionData::fontCascadeForFontUnits()

nt1m opened the pull request:
#### 5b6ab99882daefbebdbf76ce7f3174d87aea48ff
<pre>
REGRESSION(267321@main): ASSERTION FAILED: style() in WebCore::CSSToLengthConversionData::fontCascadeForFontUnits()
<a href="https://bugs.webkit.org/show_bug.cgi?id=260816">https://bugs.webkit.org/show_bug.cgi?id=260816</a>
rdar://114582938

Reviewed by Cameron McCormack.

Throw a SyntaxError when a DOMMatrix() is initialized with relative units per spec, don&apos;t let attempt to compute relative units with missing required dependencies.

<a href="https://drafts.fxtf.org/geometry/#dommatrix-parse">https://drafts.fxtf.org/geometry/#dommatrix-parse</a>

&gt; Parse transformList into parsedValue given the grammar for the CSS transform property. The result will be a &lt;transform-list&gt;, the keyword none, or failure. If parsedValue is failure, or any &lt;transform-function&gt; has &lt;length&gt; values without absolute length units, or any keyword other than none is used, then return failure. [CSS3-SYNTAX] [CSS3-TRANSFORMS]

* LayoutTests/TestExpectations
* LayoutTests/imported/w3c/web-platform-tests/css/geometry/DOMMatrix-001-expected.txt:
* Source/WebCore/css/CSSPrimitiveValue.h:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::CSSPrimitiveValue::convertingToLengthHasRequiredConversionData const):
(WebCore::CSSPrimitiveValue::convertToLength const):
(WebCore::CSSPrimitiveValue::convertingToLengthRequiresNonNullStyle const): Deleted.
* Source/WebCore/css/calc/CSSCalcExpressionNode.h:
* Source/WebCore/css/calc/CSSCalcInvertNode.h:
* Source/WebCore/css/calc/CSSCalcNegateNode.h:
* Source/WebCore/css/calc/CSSCalcOperationNode.cpp:
(WebCore::CSSCalcOperationNode::convertingToLengthHasRequiredConversionData const):
(WebCore::CSSCalcOperationNode::convertingToLengthRequiresNonNullStyle const): Deleted.
* Source/WebCore/css/calc/CSSCalcOperationNode.h:
* Source/WebCore/css/calc/CSSCalcPrimitiveValueNode.cpp:
(WebCore::CSSCalcPrimitiveValueNode::convertingToLengthHasRequiredConversionData const):
(WebCore::CSSCalcPrimitiveValueNode::convertingToLengthRequiresNonNullStyle const): Deleted.
* Source/WebCore/css/calc/CSSCalcPrimitiveValueNode.h:
* Source/WebCore/css/calc/CSSCalcValue.cpp:
(WebCore::CSSCalcValue::convertingToLengthHasRequiredConversionData const):
(WebCore::CSSCalcValue::convertingToLengthRequiresNonNullStyle const): Deleted.
* Source/WebCore/css/calc/CSSCalcValue.h:

Canonical link: <a href="https://commits.webkit.org/267378@main">https://commits.webkit.org/267378@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2e399a29158e5c6021a1cb7ffe78d97ca6f190aa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16437 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16758 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17194 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18213 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/15414 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/16628 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/19989 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16901 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/17777 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16633 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17059 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14217 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/18984 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14304 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/14901 "Passed tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15291 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15066 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19371 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15653 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/13293 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/14855 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3935 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/19224 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15478 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/css/calc/CSSCalcOperationNode.cpp:
```diff
@@ -1065,10 +1065,10 @@ void CSSCalcOperationNode::collectComputedStyleDependencies(ComputedStyleDepende
 {
-    return WTF::anyOf(m_children, [lengthConversion] (auto& child) {
-        return child->convertingToLengthRequiresNonNullStyle(lengthConversion);
+    return WTF::anyOf(m_children, [&] (auto& child) {
```
nit: keep doing the fine-grained capture like before -- `[lengthConversion, &conversionData]`

nt1m committed:
REGRESSION(267321@main): ASSERTION FAILED: style() in WebCore::CSSToLengthConversionData::fontCascadeForFontUnits()
https://bugs.webkit.org/show_bug.cgi?id=260816
rdar://114582938

Reviewed by Cameron McCormack.

Throw a SyntaxError when a DOMMatrix() is initialized with relative units per spec, don't let attempt to compute relative units with missing required dependencies.

https://drafts.fxtf.org/geometry/#dommatrix-parse

> Parse transformList into parsedValue given the grammar for the CSS transform property. The result will be a <transform-list>, the keyword none, or failure. If parsedValue is failure, or any <transform-function> has <length> values without absolute length units, or any keyword other than none is used, then return failure. [CSS3-SYNTAX] [CSS3-TRANSFORMS]

* LayoutTests/TestExpectations
* LayoutTests/imported/w3c/web-platform-tests/css/geometry/DOMMatrix-001-expected.txt:
* Source/WebCore/css/CSSPrimitiveValue.h:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::CSSPrimitiveValue::convertingToLengthHasRequiredConversionData const):
(WebCore::CSSPrimitiveValue::convertToLength const):
(WebCore::CSSPrimitiveValue::convertingToLengthRequiresNonNullStyle const): Deleted.
* Source/WebCore/css/calc/CSSCalcExpressionNode.h:
* Source/WebCore/css/calc/CSSCalcInvertNode.h:
* Source/WebCore/css/calc/CSSCalcNegateNode.h:
* Source/WebCore/css/calc/CSSCalcOperationNode.cpp:
(WebCore::CSSCalcOperationNode::convertingToLengthHasRequiredConversionData const):
(WebCore::CSSCalcOperationNode::convertingToLengthRequiresNonNullStyle const): Deleted.
* Source/WebCore/css/calc/CSSCalcOperationNode.h:
* Source/WebCore/css/calc/CSSCalcPrimitiveValueNode.cpp:
(WebCore::CSSCalcPrimitiveValueNode::convertingToLengthHasRequiredConversionData const):
(WebCore::CSSCalcPrimitiveValueNode::convertingToLengthRequiresNonNullStyle const): Deleted.
* Source/WebCore/css/calc/CSSCalcPrimitiveValueNode.h:
* Source/WebCore/css/calc/CSSCalcValue.cpp:
(WebCore::CSSCalcValue::convertingToLengthHasRequiredConversionData const):
(WebCore::CSSCalcValue::convertingToLengthRequiresNonNullStyle const): Deleted.
* Source/WebCore/css/calc/CSSCalcValue.h:

Canonical link: https://commits.webkit.org/267378@main
